PurchasesTests/Caching/DeviceCacheTests.swift

```diff
@@ -194,4 +209,117 @@ class DeviceCacheTests: XCTestCase {
+
+        expect(self.deviceCache.isPurchaserInfoCacheStale(forAppUserID: appUserID,
+                                                          isAppBackgrounded: false)) == true
+
+        let validCacheDateForForeground = Calendar.current.date(byAdding: .minute, value: -3, to: Date())!
+        self.deviceCache.setPurchaserInfoCacheTimestamp(validCacheDateForForeground, forAppUserID: appUserID)
+
+        expect(self.deviceCache.isPurchaserInfoCacheStale(forAppUserID: appUserID,
+                                                          isAppBackgrounded: true)) == false
+    }
+
+    func testIsPurchaserInfoCacheStaleForDifferentAppUserID() {
+        let mockNotificationCenter = MockNotificationCenter()
+        let otherAppUserID = "some other user"
+        let currentAppUserID = "myUser"
+        self.deviceCache = RCDeviceCache(mockUserDefaults,
+                                         offeringsCachedObject: nil,
+                                         notificationCenter: mockNotificationCenter)
+        let validCacheDate = Calendar.current.date(byAdding: .minute, value: -3, to: Date())!
+        self.deviceCache.setPurchaserInfoCacheTimestamp(validCacheDate, forAppUserID: otherAppUserID)
+
+        expect(self.deviceCache.isPurchaserInfoCacheStale(forAppUserID: currentAppUserID,
+                                                          isAppBackgrounded: true)) == true
+    }
+
+    func testIsOfferingsCacheStaleDifferentCacheLengthsForBackgroundAndForeground() {
+        let mockNotificationCenter = MockNotificationCenter()
+        let mockCachedObject = RCInMemoryCachedObject<Purchases.Offerings>()
+
+        self.deviceCache = RCDeviceCache(mockUserDefaults,
+                                         offeringsCachedObject: mockCachedObject,
+                                         notificationCenter: mockNotificationCenter)
+
+        let outdatedCacheDate = Calendar.current.date(byAdding: .hour, value: -25, to: Date())!
+        mockCachedObject.updateCacheTimestamp(with: outdatedCacheDate)
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: false)) == true
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: true)) == true
+
+        let cacheDateStaleForForeground = Calendar.current.date(byAdding: .hour, value: -23, to: Date())!
+        mockCachedObject.updateCacheTimestamp(with: cacheDateStaleForForeground)
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: false)) == true
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: true)) == false
+
+        let cacheDateValidForBoth = Calendar.current.date(byAdding: .minute, value: -3, to: Date())!
+        mockCachedObject.updateCacheTimestamp(with: cacheDateValidForBoth)
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: false)) == false
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: true)) == false
+    }
```

Comment:
I feel we need a test for the migration case where the purchaser info timestamp is nil, but the cached purchaser info is not. This can happen for people updating from an older version. I don't think it's covered in any of these tests

Reply:
good catch